contracts/extensions/BatchTradeExtension.sol

```diff
@@ -0,0 +1,198 @@
+     */
+    function initializeExtension(IDelegatedManager _delegatedManager) external onlyOwnerAndValidManager(_delegatedManager) {
+        require(_delegatedManager.isPendingExtension(address(this)), "Extension must be pending");
+
+        ISetToken setToken = _delegatedManager.setToken();
+
+        _initializeExtension(setToken, _delegatedManager);
+
+        emit BatchTradeExtensionInitialized(address(setToken), address(_delegatedManager));
+    }
+
+    /**
+     * ONLY OWNER: Initializes TradeExtension to the DelegatedManager and TradeModule to the SetToken
+     *
+     * @param _delegatedManager     Instance of the DelegatedManager to initialize
+     */
+    function initializeModuleAndExtension(IDelegatedManager _delegatedManager) external onlyOwnerAndValidManager(_delegatedManager){
+        require(_delegatedManager.isPendingExtension(address(this)), "Extension must be pending");
+
+        ISetToken setToken = _delegatedManager.setToken();
+
+        _initializeExtension(setToken, _delegatedManager);
+        _initializeModule(setToken, _delegatedManager);
+
+        emit BatchTradeExtensionInitialized(address(setToken), address(_delegatedManager));
+    }
+
+    /**
+     * ONLY MANAGER: Remove an existing SetToken and DelegatedManager tracked by the BatchTradeExtension
+     */
+    function removeExtension() external override {
+        IDelegatedManager delegatedManager = IDelegatedManager(msg.sender);
+        ISetToken setToken = delegatedManager.setToken();
+
+        _removeExtension(setToken, delegatedManager);
+    }
+
+    /**
+     * ONLY OPERATOR: Executes a batch of trades on a supported DEX. If any individual trades fail, events are emitted.
+     * @dev Although the SetToken units are passed in for the send and receive quantities, the total quantity
+     * sent and received is the quantity of SetToken units multiplied by the SetToken totalSupply.
+     *
+     * @param _setToken             Instance of the SetToken to trade
+     * @param _trades               Struct of information for individual trades
+     */
+    function batchTrade(
+        ISetToken _setToken,
+        TradeInfo[] memory _trades
+    )
+        external
+        onlyOperator(_setToken)
+        onlyAllowedAssets(_setToken, _trades)
+    {
+        for(uint256 i = 0; i < _trades.length; i++) {
+            bytes memory callData = abi.encodeWithSignature(
+                "trade(address,string,address,uint256,address,uint256,bytes)",
+                _setToken,
+                _trades[i].exchangeName,
+                _trades[i].sendToken,
+                _trades[i].sendQuantity,
+                _trades[i].receiveToken,
+                _trades[i].minReceiveQuantity,
+                _trades[i].data
+            );
+
+            try _manager(_setToken).interactManager(address(tradeModule), callData) {}
+            catch Error(string memory _err) {
+                emit StringTradeFailed(address(_setToken), i, _err);
+            }
```

Comment:
Note: here we don't `catch (bytes memory _err)` because we believe there is no path in `TradeModule.trade` that throws a `bytes` error.

Reply:
Could we just add a little comment about this in the code? e.g that everything is routed through Address.functionCallWithValue which appends a  "low-level call failed" reason string to anonymous failures from deeper in the call stack.

Reply:
Added!